<!DOCTYPE html>
<html>
<head>
  <title>Bioinformatics Bootcamp 2023</title>
  <meta charset="utf-8">
  <meta name="description" content="Bioinformatics Bootcamp 2023">
  <meta name="author" content="Dr Axel Barlow">
  <meta name="generator" content="slidify" />
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta http-equiv="X-UA-Compatible" content="chrome=1">
  <link rel="stylesheet" href="libraries/frameworks/io2012/css/default.css" media="all" >
  <link rel="stylesheet" href="libraries/frameworks/io2012/css/phone.css" 
    media="only screen and (max-device-width: 480px)" >
  <link rel="stylesheet" href="libraries/frameworks/io2012/css/slidify.css" >
  <link rel="stylesheet" href="libraries/highlighters/highlight.js/css/zenburn.css" />
  <base target="_blank"> <!-- This amazingness opens all links in a new tab. -->  <link rel=stylesheet href="./assets/css/custom.css"></link>
<link rel=stylesheet href="./assets/css/ribbons.css"></link>

  
  <!-- Grab CDN jQuery, fall back to local if offline -->
  <script src="http://ajax.aspnetcdn.com/ajax/jQuery/jquery-1.7.min.js"></script>
  <script>window.jQuery || document.write('<script src="libraries/widgets/quiz/js/jquery.js"><\/script>')</script> 
  <script data-main="libraries/frameworks/io2012/js/slides" 
    src="libraries/frameworks/io2012/js/require-1.0.8.min.js">
  </script>
  
  

</head>
<body style="opacity: 0">
  <slides class="layout-widescreen">
    
    <!-- LOGO SLIDE -->
    
  <slide class="nobackground">
    <article class="flexbox vcenter">
      <span>
        <img width='300px' src="assets/img/A1_FullColour.svg">
      </span>
    </article>
  </slide>
    <slide class="title-slide segue nobackground">
  
  <aside class="gdbar">
    <img src="assets/img/LA_Full_colour_reversed.svg">
  </aside>
  
  <hgroup class="auto-fadein">
    <h1>Bioinformatics Bootcamp 2023</h1>
    <h2>Molecular Ecology and Evolution at Bangor (MEEB)</h2>
    <p>Dr Axel Barlow<br/>email: a.barlow.@bangor.ac.uk</p>
  </hgroup>
  
    <a href="https://github.com/drabarlow/bioinformatics_bootcamp/zipball/gh-pages" class="example">
     Download
    </a>
  
  <article></article>  
      
  <footer class = 'license'>
    <a href='http://creativecommons.org/licenses/by-nc-sa/3.0/'>
    <img width = '80px' src = 'http://mirrors.creativecommons.org/presskit/buttons/88x31/png/by-nc-sa.png'>
    </a>
  </footer>
  
</slide>
    

    <!-- SLIDES -->
    <slide class="" id="slide-1" style="background:;">
  
  <article data-timings="">
    <!-- adding bold and italic options -->

<style>
em {
  font-style: italic
}
strong {
  font-weight: bold;
}
</style>

<h2>Schedule</h2>

<table><thead>
<tr>
<th>Date</th>
<th>Time</th>
<th>Activity</th>
</tr>
</thead><tbody>
<tr>
<td>Tue 03/10/2023</td>
<td>09:30 to 13:00</td>
<td>Unix-like systems and bash</td>
</tr>
<tr>
<td>Wed 04/10/2023</td>
<td>12:00 to 14:00</td>
<td>Super-computing Wales and slurm</td>
</tr>
<tr>
<td>Mon 9/10/2023</td>
<td>14:00 to 17:00</td>
<td>Illumina data and BEARCAVE</td>
</tr>
<tr>
<td>Tue 10/10/2023</td>
<td>09:30 to 17:00</td>
<td>Illumina data processing</td>
</tr>
<tr>
<td>Wed 11/10/2023</td>
<td>09:30 to 15:30</td>
<td>ANGSD, covariance and distance matrices</td>
</tr>
<tr>
<td>Fri 13/10/2023</td>
<td>14:00 to 17:00</td>
<td>Intro to R, PCA and plotting, NJ trees and plotting</td>
</tr>
</tbody></table>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-2" style="background:;">
  
  <hgroup>
    <h2>Prequisites</h2>
  </hgroup>
  
  <article data-timings="">
    <ul>
<li>a laptop that can connect to university wifi/eduroam</li>
<li>an account on super-computing Wales</li>
<li>For windows users, installation of the ssh client PuTTY</li>
<li>Basic knowledge of DNA and genome structure</li>
</ul>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-3" style="background:;">
  
  <hgroup>
    <h2>Resources</h2>
  </hgroup>
  
  <article data-timings="">
    <h3>Lecture slides</h3>

<p><a href="https://drabarlow.github.io/bioinformatics_bootcamp/">https://drabarlow.github.io/bioinformatics_bootcamp/</a></p>

<h3>Bootcamp worksheet</h3>

<p><a href="https://drabarlow.github.io/bioinformatics_bootcamp/bootcamp_worksheet.html">https://drabarlow.github.io/bioinformatics_bootcamp/bootcamp_worksheet.html</a></p>

<h3>Github repo</h3>

<p><a href="https://github.com/drabarlow/bioinformatics_bootcamp">https://github.com/drabarlow/bioinformatics_bootcamp</a></p>

<h3>Supercomputing Wales shared project</h3>

<ul>
<li>please request access to: <strong>scw2141 Bioinformatics training</strong></li>
</ul>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-4" style="background:;">
  
  <hgroup>
    <h2>Introductions</h2>
  </hgroup>
  
  <article data-timings="">
    <h3>Axel Barlow, Lecturer in Zoology at Bangor University</h3>

<p><strong>Interests</strong></p>

<ul>
<li>Population and evolutionary genomics of animals. Palaeogenomics of extinct animals. Conservation genomics of native species</li>
</ul>

<p><strong>Bioinformatics experience</strong></p>

<ul>
<li>Working in the field for ten years</li>
<li>Entirely self taught post-PhD</li>
<li>No proper bioinformatics/computer science background</li>
<li>Knowledge of bash and R.</li>
</ul>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="segue dark" id="slide-5" style="background:;">
  
  <hgroup>
    <h2>Unix-like systems and bash</h2>
  </hgroup>
  
  <article data-timings="">
    
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-6" style="background:;">
  
  <hgroup>
    <h2>Unix</h2>
  </hgroup>
  
  <article data-timings="">
    
<div style='float:left;width:48%;' class='centered'>
  <ul>
<li>Operating system developed in 1969 by Bell Labs</li>
<li>Unix philosophy: <em>operating system should provide a set of simple tools, each of which performs a limited, well-defined function.</em></li>
<li>Modular (small programs strung together)</li>
<li>Inter-process communication: &quot;pipes&quot;</li>
<li>Separate normal and &quot;super&quot; users (<code>sudo</code>)</li>
<li>Hierarchical filesystem</li>
<li>A shell for executing and combining tools</li>
<li>The basis of many subsequent OS</li>
</ul>

</div>
<div style='float:right;width:48%;'>
  <p><img src="./assets/img/Operating_system_placement.svg.png" width="40%" style="display: block; margin: auto;" /></p>

<p><img src="./assets/img/unix-an-open-group-standard.png" width="70%" style="display: block; margin: auto;" /></p>

</div>
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-7" style="background:;">
  
  <hgroup>
    <h2>Unix</h2>
  </hgroup>
  
  <article data-timings="">
    <iframe src = 'https://en.wikipedia.org/wiki/Unix' height='600px'></iframe>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-8" style="background:;">
  
  <hgroup>
    <h2>Unix-like systems</h2>
  </hgroup>
  
  <article data-timings="">
    
<div style='float:left;width:48%;' class='centered'>
  <p><strong>Mac OS</strong></p>

<ul>
<li>Released 1984</li>
<li>Developed from NeXTSTEP, which is developed from Unix</li>
<li>Proprietary, only available with Apple hardware</li>
</ul>

<p><strong>Linux</strong></p>

<ul>
<li>21 year old Linus Torvalds coded a Unix inspired OS in 1991</li>
<li>Free and open source</li>
<li>The core linux kernal available under many distributions: Ubuntu, Mint, Arch, RedHat, Android, etc.</li>
</ul>

</div>
<div style='float:right;width:48%;'>
  <p><img src="./assets/img/main-qimg-592cd3e8f514ec3fd00dd6490d889926-lq.jpeg" width="90%" style="display: block; margin: auto;" /></p>

</div>
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-9" style="background:;">
  
  <hgroup>
    <h2>MS-DOS (Windows)</h2>
  </hgroup>
  
  <article data-timings="">
    
<div style='float:left;width:66%;' class='centered'>
  <ul>
<li>Developed by Microsoft, released 1981</li>
<li>Main OS fro IMB PCs in 1980s</li>
<li>GUI introduced with Windows, released 1985</li>
<li>Largest market share (70% of PCs)</li>
<li>Some bioinformatics possible (e.g. <code>R</code> typically via <code>Rstudio</code>)</li>
<li>No <code>bash</code></li>
<li>Encoding of text files is different</li>
<li>Majority of bioinfomatics software unsupported</li>
<li>Windows subsystem for linux <a href="https://learn.microsoft.com/en-us/windows/wsl/install">https://learn.microsoft.com/en-us/windows/wsl/install</a></li>
<li>Seamless transfer between <code>DOS</code> and <code>Unix</code> not yet possible</li>
</ul>

</div>
<div style='float:right;width:30%;'>
  <p><img src="./assets/img/Msdos-icon.svg.png" width="80%" style="display: block; margin: auto;" /></p>

<p><img src="./assets/img/StartingMsdos.png" width="80%" style="display: block; margin: auto;" /></p>

</div>
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-10" style="background:;">
  
  <hgroup>
    <h2>OS comparison</h2>
  </hgroup>
  
  <article data-timings="">
    <table><thead>
<tr>
<th></th>
<th>Windows</th>
<th>Mac</th>
<th>Linux</th>
</tr>
</thead><tbody>
<tr>
<td>standard PC functions</td>
<td>yes</td>
<td>yes</td>
<td>yes</td>
</tr>
<tr>
<td>cost</td>
<td>yes</td>
<td>yes</td>
<td>free</td>
</tr>
<tr>
<td>hardware choice</td>
<td>yes</td>
<td>no</td>
<td>yes</td>
</tr>
<tr>
<td>bioinformatics</td>
<td>no</td>
<td>yes</td>
<td>yes</td>
</tr>
<tr>
<td>HPC</td>
<td>no</td>
<td>no</td>
<td>yes</td>
</tr>
<tr>
<td>open source</td>
<td>no</td>
<td>no</td>
<td>yes</td>
</tr>
<tr>
<td>active community</td>
<td>no</td>
<td>no</td>
<td>yes</td>
</tr>
<tr>
<td>games</td>
<td>yes</td>
<td>no</td>
<td>no</td>
</tr>
</tbody></table>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-11" style="background:;">
  
  <hgroup>
    <h2>Bash</h2>
  </hgroup>
  
  <article data-timings="">
    
<div style='float:left;width:48%;' class='centered'>
  <ul>
<li>a shell allows users to access OS tools</li>
<li>Unix came with the Bourne shell (<code>sh</code>), developed by Steven Bourne in 1979</li>
<li>In 1979 Brian Fox in improved version: the Bourne again shell (<code>bash</code>)</li>
<li>Most Unix-like OS use <code>bash</code> or something like it

<ul>
<li>execute standard OS functions</li>
<li>access filesystem</li>
<li>execute installed programs</li>
<li>supports bash scripts</li>
<li>pipes, auto-completion, loops, wildcards, etc.</li>
</ul></li>
</ul>

</div>
<div style='float:right;width:48%;'>
  <p><img src="./assets/img/Gnu-bash-logo.svg.png" width="90%" style="display: block; margin: auto;" /></p>

<p><img src="./assets/img/Screenshot from 2023-10-02 09-39-05.png" width="90%" style="display: block; margin: auto;" /></p>

</div>
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="segue dark" id="slide-12" style="background:;">
  
  <hgroup>
    <h2>Supercomputing Wales and slurm</h2>
  </hgroup>
  
  <article data-timings="">
    
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-13" style="background:white;">
  
  <hgroup>
    <h2>Supercomputing Wales</h2>
  </hgroup>
  
  <article data-timings="">
    <ul>
<li>£16m investment, part-funded by the European Regional Development Fund (ERDF) through Welsh Government</li>
<li>Provide university research teams access to HPC</li>
<li>Consortium of Cardiff, Swansea, Bangor and Aberystwyth</li>
<li>2 Supercomputers:

<ul>
<li>Cardiff HPC System - Hawk</li>
<li>Swansea HPC System - Sunbird</li>
</ul></li>
<li>Hawk: 280 nodes, 12,736 cores, 68.224 TB memory</li>
</ul>

<p><img src="./assets/img/lgo_supercomputingWales-1.png" width="80%" style="display: block; margin: auto;" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-14" style="background:;">
  
  <hgroup>
    <h2>Hawk</h2>
  </hgroup>
  
  <article data-timings="">
    <ul>
<li>Scientific linux OS</li>
<li><strong>compute nodes</strong>

<ul>
<li>134x Intel nodes with 20 cpus + 192 Gb RAM each</li>
<li>64x AMD nodes with 32 cpus + 256 Gb RAM each</li>
</ul></li>
<li><strong>high memory nodes</strong>

<ul>
<li>26 nodes with 384 Gb RAM each</li>
</ul></li>
<li><strong>GPU nodes</strong>

<ul>
<li>26 Nvidia P100 GPUs on 13 nodes</li>
<li>30 Nvidia V100 GPUs on 15 nodes</li>
</ul></li>
<li><strong>Storage</strong>

<ul>
<li>1192TB (usable) scratch space</li>
<li>420TB of home directory space</li>
</ul></li>
</ul>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-15" style="background:;">
  
  <hgroup>
    <h2>Hawk users</h2>
  </hgroup>
  
  <article data-timings="">
    
<div style='float:left;width:66%;' class='centered'>
  <ul>
<li>Anyone in consortium Universities</li>
<li>Guests possible, with some restrictions</li>
<li>Ability to set up shared projects</li>
<li>Generally max 10 running jobs (ways around it)</li>
<li>Three day max runtime (ways around it)</li>
<li>Generally ~50 Gb home dir</li>
<li>Generally 4 Tb scratch space (currently an issue)</li>
<li>Uses <code>Slurm</code> job scheduler</li>
<li>Many programs installed as <code>modules</code></li>
<li>No super user access</li>
<li>Can seek help from Ade Fewings (<a href="mailto:a.fewings@bangor.ac.uk">a.fewings@bangor.ac.uk</a>)</li>
</ul>

</div>
<div style='float:right;width:30%;'>
  <p><img src="./assets/img/Untitled.jpeg" width="70%" style="display: block; margin: auto;" /></p>

</div>
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-16" style="background:;">
  
  <hgroup>
    <h2>Filesystem</h2>
  </hgroup>
  
  <article data-timings="">
    <ul>
<li><code>/</code> [root] is uppermost level of filesystem</li>
<li>Everything is contained in <code>/</code></li>
<li>Directories exist within the filesystem, they can contain files and other directories</li>
<li>We specify a path through this hierarchy using forward-slashes (note back-slashes on <code>DOS</code>)</li>
<li>Our current directory is called the <code>working directory</code></li>
</ul>

<pre><code class="bash">/home/b.xlb21brx/
/scratch/b.xlb21brx/
</code></pre>

<ul>
<li>We can navigate through the filesystem (change working directory)</li>
<li>Or we can specify the patch to directories or files remotely</li>
</ul>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-17" style="background:;">
  
  <hgroup>
    <h2>Slurm</h2>
  </hgroup>
  
  <article data-timings="">
    
<div style='float:left;width:48%;' class='centered'>
  <ul>
<li>Simple Linux Utility for Resource Management: <code>Slurm</code></li>
<li>Free open source job scheduler for linux systems</li>
<li>Used on 60% of World&#39;s top 500 computers</li>
<li>Assigns user jobs to computer resources </li>
<li>Submit to queue</li>
<li>Short, low-resource jobs move faster through the queue</li>
<li>Other tools for scheduling, reporting, etc</li>
</ul>

</div>
<div style='float:right;width:48%;'>
  <p><img src="./assets/img/1024px-Slurm_logo.svg.png" width="70%" style="display: block; margin: auto;" /></p>

</div>
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="segue dark" id="slide-18" style="background:;">
  
  <hgroup>
    <h2>Illumina data</h2>
  </hgroup>
  
  <article data-timings="">
    
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-19" style="background:;">
  
  <hgroup>
    <h2>Illumina sequencing platforms</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/illumina.svg" width="100%" style="display: block; margin: auto;" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-20" style="background:;">
  
  <hgroup>
    <h2>Data output</h2>
  </hgroup>
  
  <article data-timings="">
    <table><thead>
<tr>
<th>Platform</th>
<th>Million reads</th>
<th>Read length</th>
<th>Gb data</th>
<th>Genome coverage</th>
</tr>
</thead><tbody>
<tr>
<td>MiniSeq</td>
<td>25</td>
<td>2 x 150 bp</td>
<td>7.5</td>
<td>2</td>
</tr>
<tr>
<td>MiSeq</td>
<td>25</td>
<td>2 x 300 bp</td>
<td>15</td>
<td>4</td>
</tr>
<tr>
<td>NextSeq</td>
<td>400</td>
<td>2 x 150 bp</td>
<td>120</td>
<td>33</td>
</tr>
<tr>
<td>HiSeq X</td>
<td>6000</td>
<td>2 x 150 bp</td>
<td>1800</td>
<td>500</td>
</tr>
<tr>
<td>NovaSeq</td>
<td>20000</td>
<td>2 x 150 bp</td>
<td>6000</td>
<td>1667</td>
</tr>
</tbody></table>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-21" style="background:;">
  
  <hgroup>
    <h2>Sequencing by synthesis</h2>
  </hgroup>
  
  <article data-timings="">
    <ol>
<li>Sample preparation</li>
<li>Bind DNA to flowcell, generate clusters</li>
<li>Sequencing by synthesis</li>
<li>Data analysis (in the machine)</li>
</ol>

<p><a href="https://www.youtube.com/watch?v=fCd6B5HRaZ8">https://www.youtube.com/watch?v=fCd6B5HRaZ8</a></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-22" style="background:;">
  
  <hgroup>
    <h2>Sample preparation</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/library_molecule.svg" width="100%" style="display: block; margin: auto;" /></p>

<p>*Indexes allow multiple samples to be sequenced at the same time</p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-23" style="background:;">
  
  <hgroup>
    <h2>Flow cell</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/flowcell.svg" width="80%" style="display: block; margin: auto;" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-24" style="background:white;">
  
  <hgroup>
    <h2>Cluster generation</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/cluster.svg" width="70%" style="display: block; margin: auto;" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-25" style="background:white;">
  
  <hgroup>
    <h2>Sequencing by synthesis</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/sbs.svg" width="100%" style="display: block; margin: auto;" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-26" style="background:white;">
  
  <hgroup>
    <h2>Data analysis (in the machine)</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/dataanalysis.svg" width="85%" style="display: block; margin: auto;" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-27" style="background:;">
  
  <hgroup>
    <h2>What do we sequence?</h2>
  </hgroup>
  
  <article data-timings="">
    <p>[Not an exhaustive list]</p>

<ul>
<li><strong>Whole genome sequencing (pure DNA sample from a single individual)</strong></li>
<li>Reduced representation genome data (RADseq, targeted SNPs, single individual)</li>
<li>Poolseq (multiple individuals)</li>
<li>Transcriptome (RNA sample from single tissue/individual)</li>
<li>Metabarcoding (PCR amplicon, multiple individuals/species)</li>
<li>Metagenomics (whole genomes, multiple individuals/species)</li>
</ul>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-28" style="background:;">
  
  <hgroup>
    <h2>Whole genome sequencing</h2>
  </hgroup>
  
  <article data-timings="">
    <p>Short reads from a single individual can be mapped to a reference genome assembly</p>

<p><img src="./assets/img/pseudohap.svg" width="95%" style="display: block; margin: auto;" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-29" style="background:;">
  
  <hgroup>
    <h2>Whole genome sequencing</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/Screenshot from 2022-09-08 13-58-44.png" width="100%" style="display: block; margin: auto auto auto 0;" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-30" style="background:;">
  
  <hgroup>
    <h2>Illumina summary</h2>
  </hgroup>
  
  <article data-timings="">
    <ul>
<li>The current market leader</li>
<li>Massive output</li>
<li>Many applications (genome resequencing, RADseq, transcriptomes, metabarcoding)</li>
<li>Cheap (£10 per Gb)</li>
<li>Major limitation is the read length</li>
</ul>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="segue dark" id="slide-31" style="background:;">
  
  <hgroup>
    <h2>BEARCAVE</h2>
  </hgroup>
  
  <article data-timings="">
    
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-32" style="background:;">
  
  <hgroup>
    <h2>BEARCAVE</h2>
  </hgroup>
  
  <article data-timings="">
    <ul>
<li>Nikolas Basler, Achim Klittich, Axel Barlow</li>
<li>An environment for organising, processing, and archiving Illumina data</li>
<li>BEARCAVE philosophy

<ul>
<li>All users can access all data</li>
<li>Avoid data redundancy</li>
<li>All samples processed using identical software programs and parameters</li>
<li>Incorporates sample metadata</li>
<li>Documents results of data processing</li>
<li>Easy to use wrapper scripts for programs</li>
<li>Publicly available</li>
<li>Safeguards in place to ensure consistency</li>
</ul></li>
<li>Consequently BEARCAVE is not for everyone, and has idiosyncrasies in use</li>
</ul>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-33" style="background:;">
  
  <hgroup>
    <h2>Our project: adder population genomics</h2>
  </hgroup>
  
  <article data-timings="">
    
<div style='float:left;width:48%;' class='centered'>
  <ul>
<li>Adders (<em>Vipera berus</em>) widespread across northern Eurasia</li>
<li>Threatened or near-threatened in UK</li>
<li>Illumina PE data from 18 individuals</li>
<li>5 locations</li>
<li>Data format</li>
<li>Adapter trimming and read merging</li>
<li>Map to reference genome: chr7</li>
</ul>

</div>
<div style='float:right;width:48%;'>
  <p><img src="./assets/img/male.jpg" width="90%" style="display: block; margin: auto;" /></p>

<p><img src="./assets/img/female.jpg" width="90%" style="display: block; margin: auto;" /></p>

</div>
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-34" style="background:;">
  
  <hgroup>
    <h2>Adder locations</h2>
  </hgroup>
  
  <article data-timings="">
    <table><thead>
<tr>
<th>sample</th>
<th>locality</th>
<th>colour</th>
</tr>
</thead><tbody>
<tr>
<td>adder01</td>
<td>Dublin</td>
<td>red</td>
</tr>
<tr>
<td>adder02</td>
<td>Dublin</td>
<td>red</td>
</tr>
<tr>
<td>adder03</td>
<td>Belfast</td>
<td>green</td>
</tr>
<tr>
<td>adder04</td>
<td>Belfast</td>
<td>green</td>
</tr>
<tr>
<td>adder05</td>
<td>Dublin</td>
<td>red</td>
</tr>
<tr>
<td>adder06</td>
<td>Dublin</td>
<td>red</td>
</tr>
<tr>
<td>adder07</td>
<td>Cork</td>
<td>blue</td>
</tr>
<tr>
<td>adder08</td>
<td>Cork</td>
<td>blue</td>
</tr>
<tr>
<td>adder09</td>
<td>Cork</td>
<td>blue</td>
</tr>
<tr>
<td>adder10</td>
<td>Cork</td>
<td>blue</td>
</tr>
<tr>
<td>adder11</td>
<td>Limerick</td>
<td>yellow</td>
</tr>
<tr>
<td>adder12</td>
<td>Limerick</td>
<td>yellow</td>
</tr>
<tr>
<td>adder13</td>
<td>Limerick</td>
<td>yellow</td>
</tr>
<tr>
<td>adder14</td>
<td>Limerick</td>
<td>yellow</td>
</tr>
<tr>
<td>adder15</td>
<td>Galway</td>
<td>grey</td>
</tr>
<tr>
<td>adder16</td>
<td>Galway</td>
<td>grey</td>
</tr>
<tr>
<td>adder17</td>
<td>Belfast</td>
<td>green</td>
</tr>
<tr>
<td>adder18</td>
<td>Belfast</td>
<td>green</td>
</tr>
</tbody></table>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="segue dark" id="slide-35" style="background:;">
  
  <hgroup>
    <h2>Illumina data processing</h2>
  </hgroup>
  
  <article data-timings="">
    
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-36" style="background:;">
  
  <hgroup>
    <h2>.fastq file format</h2>
  </hgroup>
  
  <article data-timings="">
    <ul>
<li>fastq is the standard output format for data from Illumina (and other) platforms</li>
</ul>

<pre><code class="bash">@A00551:758:HKTVJDSX7:4:1101:3595:6872 1:N:0:CCTGAGATGT+GGTCTAGTTG
CTGAATATGGATTTTAATTGAATCCTAAGATATTATAGCATCTTTCACTCCCTGTCCTGTGCATGTCAGA
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
</code></pre>

<ul>
<li><strong>Line 1</strong>: info on sequencer, flowcell, cluster position, indexes (sometimes)</li>
<li><strong>Line 2</strong>: called bases</li>
<li><strong>Line 3</strong>: a +</li>
<li><strong>Line 4</strong>: quality scores on Phred scale</li>
<li>10 = 90% accuracy; 20 = 99% accuracy; 30 = 99.9% accuracy</li>
<li>Recoded as single character: F = 37; ? = 30; 5 = 20; + = 10</li>
</ul>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="segue dark" id="slide-37" style="background:;">
  
  <hgroup>
    <h2>Adapter trimming and read merging</h2>
  </hgroup>
  
  <article data-timings="">
    
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-38" style="background:white;">
  
  <hgroup>
    <h2>DNA fragment length distribution</h2>
  </hgroup>
  
  <article data-timings="">
    <ul>
<li>DNA can be fragmented</li>
<li>The fragment lengths have a <strong>distribution</strong></li>
</ul>

<div style='float:left;width:48%;' class='centered'>
  <p><img src="assets/fig/unnamed-chunk-21-1.png" width="100%" height="100%" style="display: block; margin: auto;" /></p>

</div>
<div style='float:right;width:48%;'>
  <p><img src="./assets/img/kudarensis.png" width="100%" /></p>

<p>45 ka cave bear (<em>Ursus kudarensis</em>)</p>

</div>
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-39" style="background:;">
  
  <hgroup>
    <h2>Effect of insert size</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/trim1.svg" width="100%" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-40" style="background:;">
  
  <hgroup>
    <h2>Effect of insert size</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/trim2.svg" width="100%" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-41" style="background:;">
  
  <hgroup>
    <h2>Effect of insert size</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/trim3.svg" width="100%" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-42" style="background:;">
  
  <hgroup>
    <h2>Effect of insert size</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/trim4.svg" width="100%" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-43" style="background:;">
  
  <hgroup>
    <h2>Effect of insert size</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/trim5.svg" width="100%" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-44" style="background:;">
  
  <hgroup>
    <h2>Effect of insert size</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/trim6.svg" width="100%" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-45" style="background:;">
  
  <hgroup>
    <h2>Adapter trimming</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/trim9.svg" width="100%" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-46" style="background:;">
  
  <hgroup>
    <h2>Overlapping reads are merged</h2>
  </hgroup>
  
  <article data-timings="">
    <p><img src="./assets/img/trim10.svg" width="100%" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-47" style="background:;">
  
  <hgroup>
    <h2>BEARCAVE script</h2>
  </hgroup>
  
  <article data-timings="">
    <ul>
<li>decompress fastqs</li>
<li>trim adapter seqs using <code>Cutadapt</code>

<ul>
<li>30 bp min length</li>
<li>min overlap 1 bp</li>
</ul></li>
<li>merge overlapping read pair using <code>FLASH</code></li>
<li>recompress files and clean up </li>
<li>save appropriate log files</li>
</ul>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-48" style="background:;">
  
  <hgroup>
    <h2>Cutadapt</h2>
  </hgroup>
  
  <article data-timings="">
    <p><embed src="./assets/img/2011-Cutadapt_removes_adapter_sequences_from_high-throughput_sequencing_reads.pdf" width="100%" height="500" type="application/pdf" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-49" style="background:;">
  
  <hgroup>
    <h2>FLASH</h2>
  </hgroup>
  
  <article data-timings="">
    <p><embed src="./assets/img/2011-FLASH_fast_length_adjustment_of_short_reads_to_improve_genome_assemblies..pdf" width="100%" height="500" type="application/pdf" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="segue dark" id="slide-50" style="background:;">
  
  <hgroup>
    <h2>Mapping</h2>
  </hgroup>
  
  <article data-timings="">
    
  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-51" style="background:;">
  
  <hgroup>
    <h2>BEARCAVE mapping script</h2>
  </hgroup>
  
  <article data-timings="">
    <ul>
<li>decompress fastqs</li>
<li>merged (SE) and PE data processed separately

<ul>
<li>mapping using <code>bwa</code> mem algorithm</li>
<li>PCR duplicates identified and removed using <code>samtools</code></li>
<li>Reads with poor mapping quality (Q30) removed using <code>samtools</code></li>
</ul></li>
<li>SE and PE data merged</li>
<li>mapping log file generated</li>
<li>file cleanup and renaming</li>
</ul>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-52" style="background:;">
  
  <hgroup>
    <h2>bwa</h2>
  </hgroup>
  
  <article data-timings="">
    <p><embed src="./assets/img/2009-Fast_and_accurate_short_read_alignment_with_Burrows-Wheeler_transform..pdf" width="100%" height="500" type="application/pdf" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

<slide class="" id="slide-53" style="background:;">
  
  <hgroup>
    <h2>samtools (48,024 citations)</h2>
  </hgroup>
  
  <article data-timings="">
    <p><embed src="./assets/img/2009-The_Sequence_AlignmentMap_format_and_SAMtools.pdf" width="100%" height="500" type="application/pdf" /></p>

  </article>
  <!-- Presenter Notes -->
  
</slide>

    <slide class="backdrop"></slide>
  </slides>
  <div class="pagination pagination-small" id='io2012-ptoc' style="display:none;">
    <ul>
      
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=1 title='NA'>
         1
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=2 title='Prequisites'>
         2
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=3 title='Resources'>
         3
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=4 title='Introductions'>
         4
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=5 title='Unix-like systems and bash'>
         5
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=6 title='Unix'>
         6
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=7 title='Unix'>
         7
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=8 title='Unix-like systems'>
         8
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=9 title='MS-DOS (Windows)'>
         9
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=10 title='OS comparison'>
         10
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=11 title='Bash'>
         11
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=12 title='Supercomputing Wales and slurm'>
         12
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=13 title='Supercomputing Wales'>
         13
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=14 title='Hawk'>
         14
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=15 title='Hawk users'>
         15
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=16 title='Filesystem'>
         16
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=17 title='Slurm'>
         17
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=18 title='Illumina data'>
         18
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=19 title='Illumina sequencing platforms'>
         19
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=20 title='Data output'>
         20
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=21 title='Sequencing by synthesis'>
         21
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=22 title='Sample preparation'>
         22
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=23 title='Flow cell'>
         23
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=24 title='Cluster generation'>
         24
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=25 title='Sequencing by synthesis'>
         25
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=26 title='Data analysis (in the machine)'>
         26
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=27 title='What do we sequence?'>
         27
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=28 title='Whole genome sequencing'>
         28
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=29 title='Whole genome sequencing'>
         29
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=30 title='Illumina summary'>
         30
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=31 title='BEARCAVE'>
         31
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=32 title='BEARCAVE'>
         32
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=33 title='Our project: adder population genomics'>
         33
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=34 title='Adder locations'>
         34
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=35 title='Illumina data processing'>
         35
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=36 title='.fastq file format'>
         36
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=37 title='Adapter trimming and read merging'>
         37
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=38 title='DNA fragment length distribution'>
         38
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=39 title='Effect of insert size'>
         39
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=40 title='Effect of insert size'>
         40
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=41 title='Effect of insert size'>
         41
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=42 title='Effect of insert size'>
         42
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=43 title='Effect of insert size'>
         43
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=44 title='Effect of insert size'>
         44
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=45 title='Adapter trimming'>
         45
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=46 title='Overlapping reads are merged'>
         46
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=47 title='BEARCAVE script'>
         47
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=48 title='Cutadapt'>
         48
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=49 title='FLASH'>
         49
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=50 title='Mapping'>
         50
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=51 title='BEARCAVE mapping script'>
         51
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=52 title='bwa'>
         52
      </a>
    </li>
    
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=53 title='samtools (48,024 citations)'>
         53
      </a>
    </li>
    
    </ul>
  </div>  <!--[if IE]>
    <script 
      src="http://ajax.googleapis.com/ajax/libs/chrome-frame/1/CFInstall.min.js">  
    </script>
    <script>CFInstall.check({mode: 'overlay'});</script>
  <![endif]-->
</body>
  <!-- Load Javascripts for Widgets -->
  
  <!-- LOAD HIGHLIGHTER JS FILES -->
  <script src="libraries/highlighters/highlight.js/highlight.pack.js"></script>
  <script>hljs.initHighlightingOnLoad();</script>
  <!-- DONE LOADING HIGHLIGHTER JS FILES -->
   
  </html>